### PR TITLE
[IOPID-877] : use of public key to generate magic link JWE

### PR DESCRIPTION
**Pull request by pelliccm**
#### List of Changes

Use of public key to generate the magic link JWE, instead of using private key.

#### Motivation and Context

Security reasons

#### How Has This Been Tested?

Unit test, Postman

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -89,8 +89,10 @@ The following table contains the required ENV variables that the function requir
 | EXCHANGE_JWT_PRIMARY_PRIVATE_KEY| Exchange JWT Private Key (token generation)         | string|
 | EXCHANGE_JWT_ISSUER             | Exchange JWT ISSUER                                 | string|
 | EXCHANGE_JWT_TTL                | Exchange JWT TTL in seconds                         | string|
-| MAGIC_LINK_JWE_PRIMARY_PRIVATE_KEY| Magic Link JWE Private Key (token gen/validation)   | string|
-| MAGIC_LINK_JWE_SECONDARY_PRIVATE_KEY| Magic Link JWE Private Key (token gen/validation)   | string|
+| MAGIC_LINK_JWE_PRIMARY_PUB_KEY  | Magic Link JWE Public Key (token genereation)       | string|
+| MAGIC_LINK_JWE_SECONDARY_PUB_KEY| Magic Link JWE Private Key (token generation)       | string|
+| MAGIC_LINK_JWE_PRIMARY_PRIVATE_KEY| Magic Link JWE Private Key (token validation)     | string|
+| MAGIC_LINK_JWE_SECONDARY_PRIVATE_KEY| Magic Link JWE Private Key (token validation)   | string|
```
Comment (arcogabbo) [inline README.md]: this is not required as per implementation

File: README.md [hunk 2]
```diff
@@ -89,8 +89,10 @@ The following table contains the required ENV variables that the function requir
 | EXCHANGE_JWT_PRIMARY_PRIVATE_KEY| Exchange JWT Private Key (token generation)         | string|
 | EXCHANGE_JWT_ISSUER             | Exchange JWT ISSUER                                 | string|
 | EXCHANGE_JWT_TTL                | Exchange JWT TTL in seconds                         | string|
-| MAGIC_LINK_JWE_PRIMARY_PRIVATE_KEY| Magic Link JWE Private Key (token gen/validation)   | string|
-| MAGIC_LINK_JWE_SECONDARY_PRIVATE_KEY| Magic Link JWE Private Key (token gen/validation)   | string|
+| MAGIC_LINK_JWE_PRIMARY_PUB_KEY  | Magic Link JWE Public Key (token genereation)       | string|
+| MAGIC_LINK_JWE_SECONDARY_PUB_KEY| Magic Link JWE Private Key (token generation)       | string|
+| MAGIC_LINK_JWE_PRIMARY_PRIVATE_KEY| Magic Link JWE Private Key (token validation)     | string|
```
Comment (arcogabbo) [inline README.md:94]: this is not required as per implementation

**Review by arcogabbo — COMMENTED** [review]
LGTM, nice work

File: MagicLink/handler.ts [hunk 3]
```diff
@@ -32,12 +30,12 @@ type MagicLinkHandlerT = (
 
 export const magicLinkHandler = (
   issuer: NonEmptyString,
-  privateKey: NonEmptyString,
+  publicKey: NonEmptyString,
   ttl: NumberFromString,
```
Comment (BurnedMarshal) [inline MagicLink/handler.ts]: This can be a `number`

File: README.md (hunk 1 above)
Comment (BurnedMarshal) [inline README.md]: Mark the secondary keys as optional
```suggestion
| MAGIC_LINK_JWE_SECONDARY_PUB_KEY| (Optional) Magic Link JWE Private Key (token generation)       | string|
| MAGIC_LINK_JWE_PRIMARY_PRIVATE_KEY| Magic Link JWE Private Key (token validation)     | string|
| MAGIC_LINK_JWE_SECONDARY_PRIVATE_KEY| (Optional) Magic Link JWE Private Key (token validation)   | string|
```

File: utils/__tests__/jwe.test.ts [hunk 4]
```diff
@@ -1,18 +1,14 @@
 import * as E from "fp-ts/Either";
+import { exportSPKI, generateKeyPair } from "jose";
```
Comment (BurnedMarshal) [inline utils/__tests__/jwe.test.ts]: These are unused
```suggestion
```

File: utils/middlewares/__tests__/magic-link-jwe-validation-middleware.test.ts [hunk 5]
```diff
@@ -89,20 +79,15 @@ describe("MagicLinkJweValidationMiddleware", () => {
       .setIssuer(jweIssuer)
       .setIssuedAt()
       .setExpirationTime(getUnixTime(subSeconds(today, 3000)))
```
Comment (BurnedMarshal) [inline utils/middlewares/__tests__/magic-link-jwe-validation-middleware.test.ts]: Test near the validation interval
```suggestion
      .setExpirationTime(getUnixTime(subSeconds(today, 1)))
```

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: MagicLink/handler.ts [hunk 6]
```diff
@@ -53,11 +51,11 @@ export const magicLinkHandler = (
 
 export const getMagicLinkHandler = (
   issuer: NonEmptyString,
-  privateKey: NonEmptyString,
+  publicKey: NonEmptyString,
   ttl: NumberFromString,
```
Comment (BurnedMarshal) [inline MagicLink/handler.ts]:
```suggestion
  ttl: number,
```

**Review by BurnedMarshal — APPROVED** [review]
LGTM

File: MagicLink/handler.ts [hunk 7]
```diff
@@ -1,27 +1,25 @@
...
+import * as TE from "fp-ts/TaskEither";
+import { pipe } from "fp-ts/lib/function";
+
 import { ContextMiddleware } from "@pagopa/io-functions-commons/dist/src/utils/middlewares/context_middleware";
 import { RequiredBodyPayloadMiddleware } from "@pagopa/io-functions-commons/dist/src/utils/middlewares/required_body_payload";
 import {
   withRequestMiddlewares,
   wrapRequestHandler
 } from "@pagopa/io-functions-commons/dist/src/utils/request_middleware";
+import { NumberFromString } from "@pagopa/ts-commons/lib/numbers";
```
Comment (BurnedMarshal) [inline MagicLink/handler.ts]: unused
```suggestion
```

**Pull request merged**